_benchmarks/hprose_client.go

```diff
@@ -0,0 +1,168 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"reflect"
+	"runtime/pprof"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/hprose/hprose-golang/rpc"
+	"github.com/montanaflynn/stats"
+)
+
+var concurrency = flag.Int("c", 1, "concurrency")
+var total = flag.Int("n", 1, "total requests for all clients")
+var host = flag.String("s", "127.0.0.1:8972", "server ip and port")
+var cpuprofile = flag.String("cpuprofile", "", "write cpu profile to file")
+
+type RO struct {
+	Say func(args []byte) (reply []byte, err error) `simple:"true"`
+}
+
+func main() {
+	flag.Parse()
+
+	if *cpuprofile != "" {
+		f, err := os.Create(*cpuprofile)
+		if err != nil {
+			log.Fatal(err)
+		}
+		pprof.StartCPUProfile(f)
+		defer pprof.StopCPUProfile()
+	}
+
+	n := *concurrency
+	m := *total / n
+
+	fmt.Printf("concurrency: %d\nrequests per client: %d\n\n", n, m)
+
+	args := prepareArgs()
+
+	b, _ := args.Marshal()
+	fmt.Printf("message size: %d bytes\n\n", len(b))
+
+	var wg sync.WaitGroup
+	wg.Add(n * m)
+
+	var startWg sync.WaitGroup
+	startWg.Add(n)
+
+	var trans uint64
+	var transOK uint64
+
+	d := make([][]int64, n, n)
+	var ro *RO
+	client := rpc.NewTCPClient("tcp://" + *host)
+	client.SetMaxPoolSize(n)
+	client.UseService(&ro)
+
+	//it contains warmup time but we can ignore it
+	totalT := time.Now().UnixNano()
+	for i := 0; i < n; i++ {
+		dt := make([]int64, 0, m)
+		d = append(d, dt)
+
+		go func(i int) {
+			reply := &BenchmarkMessage{}
+
+			//warmup
+			ro.Say(b)
+
+			startWg.Done()
+			startWg.Wait()
+			fmt.Printf("goroutine %d started\n", i)
+
+			for j := 0; j < m; j++ {
```

Comment:
see the below comment, we want to mock real clients, each client uses one  client. There are n clients and each client sends m requests.

So should you create the client in the goroutine?